File: langs/python/src/dotlyte/watcher.py
"""File watcher for hot-reload support in DOTLYTE.

Watches configuration files for changes and triggers atomic reloads
with change diff events.
"""

from __future__ import annotations

import threading
import time
from pathlib import Path
from typing import Any, Callable

ChangeEvent = dict[str, Any]  # {"key": str, "old_value": Any, "new_value": Any}
ChangeCallback = Callable[[list[ChangeEvent]], None]
KeyChangeCallback = Callable[[ChangeEvent], None]
ErrorCallback = Callable[[Exception], None]


class ConfigWatcher:
    """Watch config files for changes, reload atomically, emit diffs.

    Args:
        files: List of file paths to watch.
        debounce_ms: Debounce time in milliseconds.

    """

    def __init__(self, files: list[str], debounce_ms: int = 100) -> None:
        self._files = [Path(f) for f in files]
        self._debounce_s = debounce_ms / 1000.0
        self._on_change: list[ChangeCallback] = []
        self._on_key_change: dict[str, list[KeyChangeCallback]] = {}
        self._on_error: list[ErrorCallback] = []
        self._running = False
        self._thread: threading.Thread | None = None
        self._reload_fn: Callable[[], dict[str, Any]] | None = None
        self._last_data: dict[str, Any] = {}
        self._last_mtimes: dict[str, float] = {}

    def start(
        self,
        reload_fn: Callable[[], dict[str, Any]],
        initial_data: dict[str, Any] | None = None,
    ) -> None:
        """Start watching files.

        Args:
            reload_fn: Function that reloads config and returns new data dict.
            initial_data: Current config data for diffing.

        """
        self._reload_fn = reload_fn
        self._last_data = initial_data or {}
        self._last_mtimes = self._get_mtimes()
        self._running = True
        self._thread = threading.Thread(target=self._poll_loop, daemon=True)
        self._thread.start()

    def on_change(self, callback: ChangeCallback) -> None:
        """Register a callback for any config change."""
        self._on_change.append(callback)

    def on_key_change(self, key: str, callback: KeyChangeCallback) -> None:
        """Register a callback for when a specific key changes."""
        if key not in self._on_key_change:
            self._on_key_change[key] = []
        self._on_key_change[key].append(callback)

    def on_error(self, callback: ErrorCallback) -> None:
        """Register a callback for reload errors."""
        self._on_error.append(callback)

    def close(self) -> None:
        """Stop watching."""
        self._running = False
        if self._thread:
            self._thread.join(timeout=2.0)
            self._thread = None

    def _poll_loop(self) -> None:
        """Poll files for mtime changes with debounce."""
        while self._running:
            time.sleep(self._debounce_s or 0.1)
            try:
                current_mtimes = self._get_mtimes()
                if current_mtimes != self._last_mtimes:
                    self._last_mtimes = current_mtimes
                    # Debounce: wait a bit more for batch writes
                    time.sleep(self._debounce_s)
                    self._handle_change()
            except Exception as e:
                for cb in self._on_error:
                    try:
                        cb(e)
                    except Exception:
                        pass

    def _handle_change(self) -> None:
        """Reload config and emit change events."""
        if not self._reload_fn:
            return

        try:
            new_data = self._reload_fn()
        except Exception as e:
            for cb in self._on_error:
                try:
                    cb(e)
                except Exception:
                    pass
            return  # Keep old config on error (atomic reload)

        events = diff_objects(self._last_data, new_data)
        self._last_data = new_data

        if events:
            for cb in self._on_change:
                try:
                    cb(events)
                except Exception:
                    pass

            for event in events:
                key = event["key"]
                if key in self._on_key_change:
                    for cb in self._on_key_change[key]:
                        try:
                            cb(event)
                        except Exception:
                            pass

    def _get_mtimes(self) -> dict[str, float]:
        """Get modification times for all watched files."""
        result: dict[str, float] = {}
        for f in self._files:
            try:
                result[str(f)] = f.stat().st_mtime
            except OSError:
                result[str(f)] = 0.0
        return result


def diff_objects(
    old: dict[str, Any],
    new: dict[str, Any],
    prefix: str = "",
) -> list[ChangeEvent]:
    """Compute the diff between two config data dicts.

    Args:
        old: Previous config data.
        new: New config data.
        prefix: Key prefix for nested diffs.

    Returns:
        List of change events.

    """
    events: list[ChangeEvent] = []
    all_keys = set(old.keys()) | set(new.keys())

    for key in all_keys:
        full_key = f"{prefix}.{key}" if prefix else key
        old_val = old.get(key)
        new_val = new.get(key)

        if isinstance(old_val, dict) and isinstance(new_val, dict):
            events.extend(diff_objects(old_val, new_val, full_key))
        elif old_val != new_val:
            events.append({
                "key": full_key,
                "old_value": old_val,
                "new_value": new_val,
            })

    return events

File: langs/ruby/lib/dotlyte/watcher.rb
# frozen_string_literal: true

require "set"

module Dotlyte
  # Polling-based file watcher for DOTLYTE v2.
  class ConfigWatcher
    # Information about a detected change.
    ChangeEvent = Struct.new(:path, :changed_keys, keyword_init: true)

    # @param files [Array<String>] files to watch
    # @param debounce_ms [Integer] polling interval in milliseconds
    def initialize(files:, debounce_ms: 300)
      @files = files
      @interval = [debounce_ms, 100].max / 1000.0
      @last_modified = {}
      @on_change = nil
      @key_watchers = {}
      @on_error = nil
      @previous_data = nil
      @running = false
      @thread = nil

      @files.each do |f|
        @last_modified[f] = File.mtime(f).to_f if File.exist?(f)
      end
    end

    # Register general change callback.
    def on_change(&block)
      @on_change = block
    end

    # Watch a specific key for changes.
    def watch_key(key, &block)
      @key_watchers[key] = block
    end

    # Register error callback.
    def on_error(&block)
      @on_error = block
    end

    # Start watching with a reload proc.
    #
    # @param reload_fn [Proc] called to reload config, must return Hash
    def start(reload_fn)
      return if @running

      @running = true
      @thread = Thread.new do
        while @running
          poll(reload_fn)
          sleep @interval
        end
      end
      @thread.abort_on_exception = false
    end

    # Stop watching.
    def stop
      @running = false
      @thread&.join(2)
      @thread = nil
    end

    private

    def poll(reload_fn)
      changed_file = nil

      @files.each do |f|
        next unless File.exist?(f)

        mtime = File.mtime(f).to_f
        prev = @last_modified[f]
        if prev.nil? || prev != mtime
          @last_modified[f] = mtime
          changed_file = f
          break
        end
      end

      return unless changed_file

      new_data = reload_fn.call
      return unless new_data

      changed_keys = if @previous_data
                       diff_maps(@previous_data, new_data)
                     else
                       flatten_keys(new_data)
                     end

      @on_change&.call(ChangeEvent.new(path: changed_file, changed_keys: changed_keys))

      if @previous_data
        changed_keys.each do |key|
          watcher = @key_watchers[key]
          next unless watcher

          old_val = Validator.get_nested(@previous_data, key)
          new_val = Validator.get_nested(new_data, key)
          watcher.call(old_val, new_val)
        end
      end

      @previous_data = new_data
    rescue StandardError => e
      @on_error&.call(e)
    end

    def diff_maps(old_map, new_map)
      old_flat = flatten_map(old_map)
      new_flat = flatten_map(new_map)
      changed = Set.new

      new_flat.each do |k, v|
        changed.add(k) if old_flat[k] != v
      end
      old_flat.each_key do |k|
        changed.add(k) unless new_flat.key?(k)
      end

      changed.to_a
    end

    def flatten_map(data, prefix = "", out = {})
      data.each do |key, value|
        full_key = prefix.empty? ? key.to_s : "#{prefix}.#{key}"
        if value.is_a?(Hash)
          flatten_map(value, full_key, out)
        else
          out[full_key] = value
        end
      end
      out
    end

    def flatten_keys(data, prefix = "", out = [])
      data.each do |key, value|
        full_key = prefix.empty? ? key.to_s : "#{prefix}.#{key}"
        if value.is_a?(Hash)
          flatten_keys(value, full_key, out)
        else
          out << full_key
        end
      end
      out
    end
  end
end
